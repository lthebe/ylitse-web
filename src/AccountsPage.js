import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Button from 'material-ui/Button';
import IconButton from 'material-ui/IconButton';
import Snackbar from 'material-ui/Snackbar';
import { withStyles } from 'material-ui/styles';
import Table, {
    TableRow, TableHead, TableCell, TableBody,
} from 'material-ui/Table';
import Typography from 'material-ui/Typography';
import CloseIcon from 'material-ui-icons/Close';
import PersonAddIcon from 'material-ui-icons/PersonAdd';

import Page from './Page';
import AccountListItem from './AccountListItem';
import ConfirmationDialog from './ConfirmationDialog';

const styles = theme => ({
    table: {
        overflowX: 'auto',
    },
    firstCell: {
        paddingLeft: 0,
    },
    lastCell: {
        '&:last-child': {
            paddingRight: [0, '!important'],
            textAlign: ['right', '!important'],
        },
    },
    button: {
        marginTop: theme.spacing.unit * 3,
        marginBottom: theme.spacing.unit * 3,
    },
    buttonIcon: {
        marginLeft: theme.spacing.unit,
    },
});

class AccountsPage extends Component {
    static propTypes = {
        classes: PropTypes.shape({
            button: PropTypes.string,
            firstCell: PropTypes.string,
            lastCell: PropTypes.string,
        }).isRequired,
    }
    constructor(props) {
        super(props);

        this.state = {
            accounts: [],
            selectedAccount: undefined,
            confirmationOpen: false,
            feedbackOpen: false,
            feedbackMessage: '',
        };
    }

    componentDidMount() {
        this.fetchAccounts();
    }

    fetchAccounts = async () => {
        try {
            const resp = await fetch('/api/accounts', {
                credentials: 'include',
                redirect: 'follow',
            });
            if (resp.redirected) {
                window.location.replace(resp.url);
            }
            const data = await resp.json();
            this.setState({
                accounts: data.resources,
            });
        } catch (e) {
            this.openFeedback(e.message);
        }
    };

    deleteAccount = async ({ id }) => {
        try {
            const resp = await fetch(`/api/accounts/${id}`, {
                method: 'DELETE',
                credentials: 'include',
                redirect: 'follow',
            });
            if (resp.redirected) {
                window.location.replace(resp.url);
            }
            const deleted = this.state.accounts.find(s => s.id === id);
            if (!resp.ok) {
                this.openFeedback(`Can't delete ${deleted.username} account`);
                this.setState({ confirmationOpen: false });
                return;
            }

            this.setState(prevState => ({
                accounts: prevState.accounts.filter(s => s.id !== id),
                confirmationOpen: false,
                feedbackOpen: true,
                feedbackMessage: `${deleted.username} account deleted`,
            }));
        } catch (e) {
            this.openFeedback(e.message);
        }
    }

    openFeedback = (feedbackMessage) => {
        this.setState({ feedbackMessage, feedbackOpen: true });
    }

    closeFeedback = () => {
        this.setState({ feedbackOpen: false });
    }

    openConfirmation = (id) => {
        this.setState(prevState => ({
            selectedAccount: prevState.accounts.find(s => s.id === id),
            confirmationOpen: true,
        }));
    }

    closeConfirmation = () => {
        this.setState({ selectedAccount: undefined, confirmationOpen: false });
    }

    render() {
        const { classes } = this.props;
        const {
            accounts, feedbackOpen, feedbackMessage, selectedAccount,
            confirmationOpen,
        } = this.state;

        return (
            <Page>
                <Typography variant="headline" component="h3">
                    Accounts
                </Typography>
                <Button
                    variant="raised"
                    color="secondary"
                    className={classes.button}
                >
                    Add new
                    <PersonAddIcon className={classes.buttonIcon} />
                </Button>
                {accounts.length > 0 &&
                    <Table className={classes.table}>
                        <TableHead>
                            <TableRow>
                                <TableCell className={classes.firstCell}>
                                    Role
                                </TableCell>
                                <TableCell>Username</TableCell>
                                <TableCell>Email</TableCell>
                                <TableCell className={classes.lastCell}>
                                    Actions
                                </TableCell>
                            </TableRow>
                        </TableHead>
                        <TableBody>
                            {accounts.map(account => (
                                <AccountListItem
                                    key={account.id}
                                    account={account}
                                    onDelete={
                                        () => this.openConfirmation(account.id)}
                                />
                            ))}
                        </TableBody>
                    </Table>
                }
                {selectedAccount &&
                    <ConfirmationDialog
                        onAction={() => this.deleteAccount(selectedAccount)}
                        onCloseConfirmation={this.closeConfirmation}
                        confirmationOpen={confirmationOpen}
                        buttonLabel="Delete"
                        label={
                            `Delete ${selectedAccount.username} account?`}
                    />
                }
                {accounts.length === 0 &&
                    <Typography variant="display3">No accounts</Typography>
                }
                <Snackbar
                    open={feedbackOpen}
                    message={feedbackMessage}
                    autoHideDuration={3000}
                    action={
                        <IconButton
                            color="inherit"
                            onClick={this.closeFeedback}
                        >
                            <CloseIcon />
                        </IconButton>
                    }
                    onClose={this.closeFeedback}
                />
            </Page>
        );
    }
}

export default withStyles(styles)(AccountsPage);
