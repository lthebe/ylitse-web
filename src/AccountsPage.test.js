import React from 'react';
import { shallow } from 'enzyme';

import AccountsPage from './AccountsPage';

describe('AccountsPage', () => {
    test('renders correctly', () => {
        const wrapper = shallow(<AccountsPage />).dive();

        expect(wrapper).toMatchSnapshot();
    });

    test('open and close feedback while fetching accounts', async () => {
        const wrapper = shallow(<AccountsPage />).dive();
        const error = 'Can not fetch accounts';
        window.fetch = jest.fn();
        window.fetch.mockReturnValue({
            json: () => {
                throw new Error(error);
            },
        });
        await wrapper.instance().fetchAccounts();
        wrapper.update();
        expect(wrapper.state('feedbackOpen')).toBe(true);
        expect(wrapper.find({ message: error, open: true })).toHaveLength(1);
        wrapper.instance().closeFeedback();
        wrapper.update();
        expect(wrapper.state('feedbackOpen')).toBe(false);
        expect(wrapper.find({ message: error, open: false })).toHaveLength(1);
    });

    test('accounts are fetched', async () => {
        const wrapper = shallow(<AccountsPage />).dive();
        const accounts = [{
            id: 23333,
            username: 'admin',
            role: 'admin',
            email: 'admin@admin.fi',
        },
        {
            id: 23333,
            username: 'admin',
            role: 'admin',
            email: 'admin@admin.fi',
        },
        ];
        const res = new Response(`{"resources": ${JSON.stringify(accounts)}}`, {
            status: 200,
            headers: {
                'content-type': 'application/json',
            },
        });

        window.fetch = jest.fn();
        window.fetch.mockReturnValue(Promise.resolve(res));
        await wrapper.instance().fetchAccounts();
        wrapper.update();
        expect(wrapper.state('accounts'))
            .toEqual(expect.arrayContaining(accounts));
    });
});
