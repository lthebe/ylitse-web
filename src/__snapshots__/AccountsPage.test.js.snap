// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsPage renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AccountsPage
    classes={
      Object {
        "button": "AccountsPage-button-4",
        "buttonIcon": "AccountsPage-buttonIcon-5",
        "firstCell": "AccountsPage-firstCell-2",
        "lastCell": "AccountsPage-lastCell-3",
        "table": "AccountsPage-table-1",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <WithStyles(Typography)
          component="h3"
          variant="headline"
        >
          Accounts
        </WithStyles(Typography)>,
        <WithStyles(Button)
          className="AccountsPage-button-4"
          color="secondary"
          variant="raised"
        >
          Add new
          <pure(PersonAdd)
            className="AccountsPage-buttonIcon-5"
          />
        </WithStyles(Button)>,
        false,
        undefined,
        <WithStyles(Typography)
          variant="display3"
        >
          No accounts
        </WithStyles(Typography)>,
        <WithStyles(Snackbar)
          action={
            <WithStyles(IconButton)
              color="inherit"
              onClick={[Function]}
            >
              <pure(Close) />
            </WithStyles(IconButton)>
          }
          autoHideDuration={3000}
          message=""
          onClose={[Function]}
          open={false}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Accounts",
          "component": "h3",
          "variant": "headline",
        },
        "ref": null,
        "rendered": "Accounts",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            "Add new",
            <pure(PersonAdd)
              className="AccountsPage-buttonIcon-5"
            />,
          ],
          "className": "AccountsPage-button-4",
          "color": "secondary",
          "variant": "raised",
        },
        "ref": null,
        "rendered": Array [
          "Add new",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "className": "AccountsPage-buttonIcon-5",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      false,
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "No accounts",
          "variant": "display3",
        },
        "ref": null,
        "rendered": "No accounts",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "action": <WithStyles(IconButton)
            color="inherit"
            onClick={[Function]}
          >
            <pure(Close) />
          </WithStyles(IconButton)>,
          "autoHideDuration": 3000,
          "message": "",
          "onClose": [Function],
          "open": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <WithStyles(Typography)
            component="h3"
            variant="headline"
          >
            Accounts
          </WithStyles(Typography)>,
          <WithStyles(Button)
            className="AccountsPage-button-4"
            color="secondary"
            variant="raised"
          >
            Add new
            <pure(PersonAdd)
              className="AccountsPage-buttonIcon-5"
            />
          </WithStyles(Button)>,
          false,
          undefined,
          <WithStyles(Typography)
            variant="display3"
          >
            No accounts
          </WithStyles(Typography)>,
          <WithStyles(Snackbar)
            action={
              <WithStyles(IconButton)
                color="inherit"
                onClick={[Function]}
              >
                <pure(Close) />
              </WithStyles(IconButton)>
            }
            autoHideDuration={3000}
            message=""
            onClose={[Function]}
            open={false}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Accounts",
            "component": "h3",
            "variant": "headline",
          },
          "ref": null,
          "rendered": "Accounts",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              "Add new",
              <pure(PersonAdd)
                className="AccountsPage-buttonIcon-5"
              />,
            ],
            "className": "AccountsPage-button-4",
            "color": "secondary",
            "variant": "raised",
          },
          "ref": null,
          "rendered": Array [
            "Add new",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "className": "AccountsPage-buttonIcon-5",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        false,
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "No accounts",
            "variant": "display3",
          },
          "ref": null,
          "rendered": "No accounts",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "action": <WithStyles(IconButton)
              color="inherit"
              onClick={[Function]}
            >
              <pure(Close) />
            </WithStyles(IconButton)>,
            "autoHideDuration": 3000,
            "message": "",
            "onClose": [Function],
            "open": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
